src/modules/commander/failsafe/failsafe.cpp

```diff
@@ -0,0 +1,531 @@
+		case LowBatteryAction::Return:
+			options.action = Action::RTL;
+			break;
+
+		case LowBatteryAction::ReturnOrLand:
+		case LowBatteryAction::Land:
+			options.action = Action::Land;
+			break;
+
+		case LowBatteryAction::Warning:
+			options.action = Action::Warn;
+			break;
+		}
+
+		break;
+	}
+
+	return options;
+}
+
+FailsafeBase::ActionOptions Failsafe::fromQuadchuteActParam(int param_value)
+{
+	ActionOptions options{};
+
+	switch (param_value) {
+	case -1:
+	default:
+		options.action = Action::Warn;
+		break;
+
+	case 0:
+		options.action = Action::RTL;
+		options.clear_condition = ClearCondition::OnModeChangeOrDisarm;
+		break;
+
+	case 1:
+		options.action = Action::Land;
+		options.clear_condition = ClearCondition::OnModeChangeOrDisarm;
+		break;
+
+	case 2:
+		options.action = Action::Hold;
+		options.clear_condition = ClearCondition::OnModeChangeOrDisarm;
+		break;
+	}
+
+	return options;
+}
+
+FailsafeBase::Action Failsafe::fromOffboardLossActParam(int param_value, uint8_t &user_intended_mode)
+{
+	Action action{Action::None};
+
+	switch (param_value) {
+	case 0:
+	default:
+		action = Action::FallbackPosCtrl;
+		user_intended_mode = vehicle_status_s::NAVIGATION_STATE_POSCTL;
+		break;
+
+	case 1:
+		action = Action::FallbackAltCtrl;
+		user_intended_mode = vehicle_status_s::NAVIGATION_STATE_ALTCTL;
+		break;
+
+	case 2:
+		action = Action::FallbackStab;
+		user_intended_mode = vehicle_status_s::NAVIGATION_STATE_STAB;
+		break;
+
+	case 3:
+		action = Action::RTL;
+		user_intended_mode = vehicle_status_s::NAVIGATION_STATE_AUTO_RTL;
+		break;
+
+	case 4:
+		action = Action::Land;
+		user_intended_mode = vehicle_status_s::NAVIGATION_STATE_AUTO_LAND;
+		break;
+
+	case 5:
+		action = Action::Hold;
+		user_intended_mode = vehicle_status_s::NAVIGATION_STATE_AUTO_LOITER;
+		break;
+
+	case 6:
+		action = Action::Terminate;
+		user_intended_mode = vehicle_status_s::NAVIGATION_STATE_TERMINATION;
+		break;
+
+	case 7:
+		action = Action::Disarm;
+		break;
+	}
+
+	return action;
+}
+
+void Failsafe::checkStateAndMode(const hrt_abstime &time_us, const State &state,
+				 const vehicle_status_flags_s &status_flags)
+{
+	updateArmingState(time_us, state.armed, status_flags);
+
+	const bool in_forward_flight = state.vehicle_type == vehicle_status_s::VEHICLE_TYPE_FIXED_WING
+				       || state.vtol_in_transition_mode;
+
+	// Do not enter failsafe while doing a vtol takeoff after the vehicle has started a transition and before it reaches the loiter
+	// altitude. The vtol takeoff navigaton mode will set mission_finished to true as soon as the loiter is established
+	const bool ignore_link_failsafe = state.user_intended_mode == vehicle_status_s::NAVIGATION_STATE_AUTO_VTOL_TAKEOFF
+					  && in_forward_flight && !state.mission_finished;
+
+	// RC loss
+	if (!status_flags.rc_signal_lost) {
+		// If RC was lost and arming was allowed, consider it optional until we regain RC
+		_rc_lost_at_arming = false;
```

Comment:
Why do we need this special case for RC lost at arming?

Reply:
To match the existing behavior. RC/manual control should later become a mode requirement, possibly with some parameters whether RC is required in general.
As this PR is mostly a refactoring, I did not want to change the existing behavior too much, as I see various use-cases with different behaviors. So we can (and should) change that separately.
I kind of also wanted to convince myself all the existing logic is possible within the new framework.